metadata/src/main/java/org/apache/kafka/controller/PartitionChangeBuilder.java

```diff
@@ -123,6 +127,7 @@ public PartitionChangeBuilder(
         this.targetElr = Replicas.toList(partition.elr);
         this.targetLastKnownElr = Replicas.toList(partition.lastKnownElr);
         this.targetLeaderRecoveryState = partition.leaderRecoveryState;
+        this.targetDirectories = DirectoryId.createAssignmentMap(partition.replicas, partition.directories);
```

Comment:
One thing I'm trying to understand is why we should use a map here rather than a list of the same length as the replicas list. This also applies to other uses of `createAssignmentMap` -- is there ever a place where we would be better off with a map rather than just an array of the right size?

Reply:
What should `.setDirectory(brokerId, dirId)` do if `brokerId` is not in the replica list yet?

I don't feel good about enforcing order in the builder pattern.

Why is a map bad here?

Reply:
That's a fair point. A map lets you set the directories before you set the replica information. So never mind.

I do still think a map is a bad representation elsewhere (like internally in `PartitionRegistration.java`) but we can talk about that elsewhere